### PR TITLE
plan: first step of join auto-select.

### PR DESCRIPTION
**Pull request by hanfei1991**
We consider different types of join execution in cbo framework. But we should still refine the cost calculation in the near future.

@zimulala @winoros @lamxTyler PTAL

### CODE REVIEW

**Comment by shenli** [comment]
Please make sure that you can pass tidb-test.

File: executor/distsql.go [hunk 1]
```diff
@@ -165,8 +165,10 @@ func indexValuesToKVRanges(tid, idxID int64, values [][]types.Datum) ([]kv.KeyRa
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		rangeKey := tablecodec.EncodeIndexSeekKey(tid, idxID, valKey)
-		krs = append(krs, kv.KeyRange{StartKey: rangeKey, EndKey: rangeKey})
+		valKeyNext := []byte(kv.Key(valKey).PrefixNext())
```
Comment (shenli) [inline executor/distsql.go]: This is a bug fix? Should we add a test case and cherry-pick it to rc3?

File: plan/logical_plan_builder.go [hunk 2]
```diff
@@ -1148,7 +1148,9 @@ func (b *planBuilder) buildApplyWithJoinType(outerPlan, innerPlan LogicalPlan, t
 	b.optFlag = b.optFlag | flagBuildKeyInfo
 	b.optFlag = b.optFlag | flagDecorrelate
 	ap := LogicalApply{LogicalJoin: LogicalJoin{JoinType: tp}}.init(b.allocator, b.ctx)
-
+	if tp == LeftOuterJoin {
```
Comment (shenli) [inline plan/logical_plan_builder.go]: Consider RightOuterJoin?

File: plan/new_physical_plan_builder.go [hunk 3]
```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
-	for {
-		switch x := innerChild.(type) {
-		case *DataSource:
-			indices, includeTableScan := availableIndices(x.indexHints, x.tableInfo)
-			if includeTableScan && len(innerJoinKeys) == 1 {
-				pkCol := x.getPKIsHandleCol()
-				if pkCol != nil && innerJoinKeys[0].Equal(pkCol, nil) {
-					useTableScan = true
-				}
+	if x, ok := innerChild.(*DataSource); ok {
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Why not consider Selection? If we only consider DataSource there, it is better to write code like this:
x, ok := innerChild.(*DataSource)
if !ok {
return nil
}

File: plan/new_physical_plan_builder.go [hunk 4]
```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
+	return nil
+}
+
+func (p *PhysicalIndexJoin) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
+	if !prop.isEmpty() {
+		return nil
+	}
+	requiredProps1 := make([]*requiredProp, 2)
+	requiredProps1[p.outerIndex] = &requiredProp{taskTp: rootTaskType}
+	requiredProps1[1-p.outerIndex] = &requiredProp{taskTp: copSingleReadTaskType, cols: p.InnerJoinKeys}
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Why this is copSingleReadTaskType? And why line 184 use copDoubleReadTaskType?

File: plan/new_physical_plan_builder.go [hunk 5]
```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
-	}.init(p.allocator, p.ctx, p.children[outerIdx], p.children[1-outerIdx])
-	task := join.attach2Task(outerTask, innerTask)
-	task = prop.enforceProperty(task, p.ctx, p.allocator)
-	return task, nil
+	return nil
+}
+
+func (p *PhysicalIndexJoin) getChildrenPossibleProps(prop *requiredProp) [][]*requiredProp {
+	if !prop.isEmpty() {
+		return nil
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Why returns nil here?

File: plan/new_physical_plan_builder.go [hunk 6]
```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
+		if prop.equal(lProp) && p.JoinType == RightOuterJoin {
+			return nil
+		}
+	}
+
+	return [][]*requiredProp{{lProp, rProp}}
 }
 
-// tryToGetIndexJoin tries to get index join plan. If fails, it returns nil.
 // Currently we only check by hint. If we prefer the left index join but the join type is right outer, it will fail to return.
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: Should we update this comment?

> Currently we only check by hint.

File: plan/new_physical_plan_builder.go [hunk 7]
```diff
@@ -101,172 +101,185 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
-func (p *LogicalJoin) tryToGetIndexJoin(prop *requiredProp) (bestTask task, err error) {
+func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
```
Comment (shenli) [inline plan/new_physical_plan_builder.go]: What is the second return value used for?

File: executor/distsql.go (hunk 1 above)
Comment (hanfei1991) [inline executor/distsql.go]: No need, it is only used by new planner.

File: plan/logical_plan_builder.go (hunk 2 above)
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: No need, apply never use a right join

File: plan/new_physical_plan_builder.go (hunk 3 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Consider selection bring more complexity with auto-select, which may bring unexpected probs.

File: plan/new_physical_plan_builder.go (hunk 5 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Batch Index Join never pass any order.

File: plan/new_physical_plan_builder.go (hunk 4 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: We don't know the type of index reading, so we should try both

File: plan/new_physical_plan_builder.go (hunk 6 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: yeah

File: plan/new_physical_plan_builder.go (hunk 7 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: If it is selected by hint, we force to choose this join.

Comment (shenli) [inline plan/new_physical_plan_builder.go]: OK, please comment it.

File: plan/new_physical_plan_builder.go (hunk 4 above)
Comment (shenli) [inline plan/new_physical_plan_builder.go]: But the cost for singleRead and doubleRead is not the same. What if we choose the first prop but   need to do douleRead indeed?

Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: we will choose a plan with min cost.

Comment (shenli) [inline plan/new_physical_plan_builder.go]: The prop with singleRead always has less cost than the one with doubleRead?

Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: If it cannot do single read, it will generate a invalid task that never be choosed.

Comment (shenli) [inline plan/new_physical_plan_builder.go]: Got it.

Comment (shenli) [inline plan/new_physical_plan_builder.go]: We should also comment it here.

File: plan/new_physical_plan_builder.go [hunk 8]
```diff
@@ -101,172 +101,186 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
+		OtherConditions: p.OtherConditions,
+		Outer:           p.JoinType != InnerJoin,
+		OuterJoinKeys:   outerJoinKeys,
+		InnerJoinKeys:   innerJoinKeys,
+		DefaultValues:   p.DefaultValues,
+	}.init(p.allocator, p.ctx, p.children[outerIdx], p.children[1-outerIdx])
+	return join
+}
+
 // convertToIndexJoin will generate index join by required properties and outerIndex. OuterIdx points out the outer child,
```
Comment (lamxTyler) [inline plan/new_physical_plan_builder.go]: The comment should be updated.

File: plan/new_physical_plan_builder.go [hunk 9]
```diff
@@ -311,24 +303,10 @@ func (p *LogicalJoin) convert2SemiJoin(prop *requiredProp) (task, error) {
...
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	rTask, err := rChild.convert2NewPhysicalPlan(&requiredProp{taskTp: rootTaskType})
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	task := semiJoin.attach2Task(lTask, rTask)
-	// Because hash join is executed by multiple goroutines, it will not propagate physical property any more.
-	// TODO: We will consider the problem of property again for parallel execution.
```
Comment (winoros) [inline plan/new_physical_plan_builder.go]: Is this `TODO` done?

File: plan/new_physical_plan_builder.go [hunk 10]
```diff
@@ -101,172 +101,186 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
-			task, err = p.convert2MergeJoin(prop)
-		} else if task, err = p.tryToGetIndexJoin(prop); task == nil && err == nil {
-			task, err = p.convert2HashJoin(prop)
+			return []PhysicalPlan{mj}
 		}
+		joins := make([]PhysicalPlan, 0, 5)
+		if len(p.EqualConditions) == 1 {
+			joins = append(joins, mj)
+		}
+		idxJoins, forced := p.tryToGetIndexJoin()
```
Comment (winoros) [inline plan/new_physical_plan_builder.go]: If we fail to create index join such that the hint is used in somewhere no child of join is datasource, should we go back to create hash join?

File: executor/join_test.go [hunk 11]
```diff
@@ -470,7 +470,7 @@ func (s *testSuite) TestSubquery(c *C) {
 	tk.MustExec("create table t(id int primary key, v int)")
 	tk.MustExec("insert into t values(1, 1), (2, 2), (3, 3)")
 	result = tk.MustQuery("select (select t.id from t where s.id < 2 and t.id = s.id) from t s")
-	result.Check(testkit.Rows("1", "<nil>", "<nil>"))
+	result.Sort().Check(testkit.Rows("1", "<nil>", "<nil>"))
```
Comment (hanfei1991) [inline executor/join_test.go]: This join is choosed as index join, which returns <nil> firstly.

File: plan/new_physical_plan_builder.go (hunk 10 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: Yes. So If the `forced` is false, we will append HashJoin to possible physical plans.

File: plan/new_physical_plan_builder.go (hunk 9 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: This TODO comment will be added again when I begin to refactor the logic of cost calculation.

Comment (winoros) [inline plan/new_physical_plan_builder.go]: okay.

**Comment by hanfei1991** [comment]
After adding some comments, the code is more readable.
PTAL @shenli @winoros @lamxTyler

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@winoros @lamxTyler PTAL

File: plan/new_physical_plan_builder.go [hunk 12]
```diff
@@ -101,172 +101,191 @@ func joinKeysMatchIndex(keys []*expression.Column, index *model.IndexInfo) []int
...
+// tryToGetIndexJoin will get index join by hints. If we can generate a valid index join by hint, the second return value
+// will be true, which means we are forced to choose this index join. Otherwise we will select a join algorithm with min-cost.
+func (p *LogicalJoin) tryToGetIndexJoin() ([]PhysicalPlan, bool) {
+	if len(p.EqualConditions) == 0 {
+		return nil, false
+	}
+	plans := make([]PhysicalPlan, 0, 2)
 	leftOuter := (p.preferINLJ & preferLeftAsOuter) > 0
 	if leftOuter {
 		if p.JoinType != RightOuterJoin {
```
Comment (zimulala) [inline plan/new_physical_plan_builder.go]: Merge the two `if` sentences.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**